File: android/app/src/main/java/com/app/matter/ESPMatterControl.kt
/*
 * SPDX-FileCopyrightText: 2026 Espressif Systems (Shanghai) CO LTD
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.app.matter

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import chip.devicecontroller.ChipDeviceController
import chip.devicecontroller.GetConnectedDeviceCallbackJni.GetConnectedDeviceCallback
import chip.devicecontroller.InvokeCallback
import chip.devicecontroller.ReportCallback
import chip.devicecontroller.ResubscriptionAttemptCallback
import chip.devicecontroller.SubscriptionEstablishedCallback
import chip.devicecontroller.WriteAttributesCallback
import chip.devicecontroller.model.AttributeWriteRequest
import chip.devicecontroller.model.ChipAttributePath
import chip.devicecontroller.model.ChipEventPath
import chip.devicecontroller.model.ChipPathId
import chip.devicecontroller.model.InvokeElement
import chip.devicecontroller.model.NodeState
import chip.devicecontroller.model.Status
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.math.BigInteger
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * Native Matter control adapter for Android.
 *
 * Exposes the four canonical Matter Interaction Model operations —
 * `read`, `write`, `invoke`, `subscribe` (plus their lifecycle siblings
 * `init`, `shutdown`, `unsubscribe`). No semantic translation lives
 * here: callers send raw cluster / attribute / command ids and Matter
 * data-value (`MTRDataValueDictionary`) payloads; this layer encodes
 * them to CHIP TLV via [MatterDataValueCodec], dispatches via
 * [ChipDeviceController], and translates inbound reports back to JS
 * primitives over the `ESPMatter:attributeReport` device event.
 *
 * Cluster-specific semantics (semantic units → Matter command, mode
 * pickers, on/off bool → On/Off cmd, etc.) belong above this layer —
 * either in TypeScript panels/hooks or in the Matter SDK's outbound
 * transformer.
 *
 * Lifecycle: reuses [FabricSessionManager]'s active [ChipClient] when
 * one is in memory (typical right after commissioning). On cold start,
 * lazily reconstructs the [ChipClient] from stored [FabricInfo]. One
 * controller serves all Matter nodes on the user's fabric.
 */
class ESPMatterControl(private val reactContext: ReactApplicationContext) {

    companion object {
        private const val TAG = "ESPMatterControl"
        private const val DEFAULT_TIMEOUT_MS = 15000
        private const val EVENT_ATTRIBUTE_REPORT = "ESPMatter:attributeReport"
        private const val DEFAULT_MIN_INTERVAL_SEC = 1
        private const val DEFAULT_MAX_INTERVAL_SEC = 30
    }

    private val mainHandler = Handler(Looper.getMainLooper())

    /** Cached connected-device pointers, keyed by uint64 matter node id. */
    private val devicePtrs = ConcurrentHashMap<Long, Long>()

    private data class SubscriptionEntry(
        val handle: String,
        val matterNodeId: Long,
        @Volatile var subscriptionId: Long = 0L
    )

    private val subscriptions = ConcurrentHashMap<String, SubscriptionEntry>()
    private val subscriptionHandleSeq = AtomicLong(0L)

    // -----------------------------------------------------------------------
    // OPTIONAL LIFECYCLE
    // -----------------------------------------------------------------------

    /** No-op today; the active [ChipClient] is initialised by commissioning. */
    fun init(@Suppress("UNUSED_PARAMETER") config: ReadableMap?, promise: Promise) {
        promise.resolve(successResult())
    }

    /** Tears down all device pointers and active subscriptions. */
    fun shutdown(promise: Promise) {
        try {
            for (entry in subscriptions.values) {
                shutdownSubscriptionEntry(entry)
            }
            subscriptions.clear()
            devicePtrs.clear()
            promise.resolve(successResult())
        } catch (e: Exception) {
            promise.reject("SHUTDOWN_FAILED", e.message, e)
        }
    }

    // -----------------------------------------------------------------------
    // READ
    // -----------------------------------------------------------------------

    fun read(
        matterNodeIdStr: String,
        endpoint: Int,
        clusterId: Double,
        attributeId: Double,
        promise: Promise
    ) {
        val nodeId = parseMatterNodeId(matterNodeIdStr) ?: run {
            promise.reject("INVALID_ARG", "Invalid matterNodeId"); return
        }
        val controller = runCatching { ensureChipClient().chipDeviceController }
            .getOrElse { promise.reject("NO_FABRIC", it.message, it); return }

        // Hex tags help correlate with the SDK's read() call sites and with
        // logcat lines emitted by the Matter SDK / panel hooks.
        val readTag = "read[ep=$endpoint,clu=0x${clusterId.toLong().toString(16)}," +
                "att=0x${attributeId.toLong().toString(16)}]"
        Log.d(TAG, "$readTag → request")
        getConnectedDevicePointer(nodeId,
            onSuccess = { devicePtr ->
                val path = ChipAttributePath.newInstance(
                    endpoint, clusterId.toLong(), attributeId.toLong()
                )
                val callback = object : ReportCallback {
                    override fun onReport(nodeState: NodeState?) {
                        try {
                            val state = nodeState
                                ?.getEndpointState(endpoint)
                                ?.getClusterState(clusterId.toLong())
                                ?.getAttributeState(attributeId.toLong())
                            val raw = state?.value
                            // Raw-shape line: tells us if CHIP delivered
                            // null, a primitive, a Map<Integer,Object>
                            // (TLV struct) or a ChipStructs$* data class.
                            // Critical when diagnosing why the SDK
                            // resolver gets `undefined` for an attribute.
                            Log.d(
                                TAG,
                                "$readTag ← raw=${if (raw == null) "null" else MatterDataValueCodec.previewValue(raw)}"
                            )
                            val jsValue = MatterDataValueCodec.attributeValueToJs(raw)
                            Log.d(
                                TAG,
                                "$readTag ← decoded type=${jsValue?.javaClass?.simpleName ?: "null"}"
                            )
                            val result = Arguments.createMap().apply {
                                putBoolean("success", true)
                                putAnyValue("value", jsValue)
                            }
                            promise.resolve(result)
                        } catch (e: Exception) {
                            Log.e(TAG, "$readTag processing failed: ${e.message}", e)
                            promise.reject("READ_FAILED", e.message, e)
                        }
                    }

                    override fun onError(
                        attributePath: ChipAttributePath?,
                        eventPath: ChipEventPath?,
                        ex: Exception
                    ) {
                        Log.w(TAG, "$readTag failed: ${ex.message}")
                        promise.reject("READ_FAILED", ex.message, ex)
                    }
                }
                try {
                    controller.readAttributePath(
                        callback, devicePtr, listOf(path), DEFAULT_TIMEOUT_MS
                    )
                } catch (e: Exception) {
                    promise.reject("READ_FAILED", e.message, e)
                }
            },
            onFailure = { e -> promise.reject("CONNECT_FAILED", e.message, e) }
        )
    }

    // -----------------------------------------------------------------------
    // WRITE (raw Matter attribute write — value is `MatterDataValue` shape)
    // -----------------------------------------------------------------------

    fun write(
        matterNodeIdStr: String,
        endpoint: Int,
        clusterId: Double,
        attributeId: Double,
        value: ReadableMap?,
        promise: Promise
    ) {
        val nodeId = parseMatterNodeId(matterNodeIdStr) ?: run {
            promise.reject("INVALID_ARG", "Invalid matterNodeId"); return
        }
        val controller = runCatching { ensureChipClient().chipDeviceController }
            .getOrElse { promise.reject("NO_FABRIC", it.message, it); return }

        val tlv = try {
            MatterDataValueCodec.encodeToTlv(value)
        } catch (e: Exception) {
            promise.reject("ENCODE_FAILED", "MatterDataValue encode failed: ${e.message}", e)
            return
        }

        getConnectedDevicePointer(nodeId,
            onSuccess = { devicePtr ->
                val request = AttributeWriteRequest.newInstance(
                    ChipPathId.forId(endpoint.toLong()),
                    ChipPathId.forId(clusterId.toLong()),
                    ChipPathId.forId(attributeId.toLong()),
                    tlv
                )
                val callback = object : WriteAttributesCallback {
                    override fun onResponse(attributePath: ChipAttributePath?, status: Status?) {
                        promise.resolve(successResult())
                    }
                    override fun onError(attributePath: ChipAttributePath?, ex: Exception?) {
                        Log.w(TAG, "write failed: ${ex?.message}")
                        promise.reject("WRITE_FAILED", ex?.message ?: "write failed", ex)
                    }
                }
                try {
                    controller.write(callback, devicePtr, listOf(request), 0, 0)
                } catch (e: Exception) {
                    promise.reject("WRITE_FAILED", e.message, e)
                }
            },
            onFailure = { e -> promise.reject("CONNECT_FAILED", e.message, e) }
        )
    }

    // -----------------------------------------------------------------------
    // INVOKE (raw Matter command — payload is `MatterDataValue` Structure)
    // -----------------------------------------------------------------------

    fun invoke(
        matterNodeIdStr: String,
        endpoint: Int,
        clusterId: Double,
        commandId: Double,
        commandFields: ReadableMap?,
        promise: Promise
    ) {
        invokeInternal(matterNodeIdStr, endpoint, clusterId.toLong(), commandId.toLong(), commandFields,
            onSuccess = { promise.resolve(successResult()) },
            onFailure = { code, msg, t -> promise.reject(code, msg, t) }
        )
    }

    fun encodeCommandFieldsToTlvHex(commandFields: ReadableMap?, promise: Promise) {
        try {
            val tlv = MatterDataValueCodec.encodeCommandFieldsToTlv(commandFields)
            val hex = tlv.joinToString("") { b -> "%02x".format(b) }
            promise.resolve("0x$hex")
        } catch (e: Exception) {
            promise.reject("ENCODE_FAILED", e.message, e)
        }
    }

    private fun invokeInternal(
        matterNodeIdStr: String,
        endpoint: Int,
        clusterId: Long,
        commandId: Long,
        commandFields: ReadableMap?,
        onSuccess: () -> Unit,
        onFailure: (String, String, Throwable?) -> Unit
    ) {
        val nodeId = parseMatterNodeId(matterNodeIdStr) ?: run {
            onFailure("INVALID_ARG", "Invalid matterNodeId", null); return
        }
        val controller = runCatching { ensureChipClient().chipDeviceController }
            .getOrElse { onFailure("NO_FABRIC", it.message ?: "no fabric", it); return }

        val tlv = try {
            MatterDataValueCodec.encodeCommandFieldsToTlv(commandFields)
        } catch (e: Exception) {
            onFailure("ENCODE_FAILED", "MatterDataValue encode failed: ${e.message}", e); return
        }

        getConnectedDevicePointer(nodeId,
            onSuccess = { devicePtr ->
                val element = InvokeElement.newInstance(endpoint, clusterId, commandId, tlv, null)
                val callback = object : InvokeCallback {
                    override fun onResponse(invokeElement: InvokeElement?, successCode: Long) {
                        onSuccess()
                    }
                    override fun onError(ex: Exception?) {
                        Log.w(TAG, "invoke failed: ${ex?.message}")
                        onFailure("INVOKE_FAILED", ex?.message ?: "invoke failed", ex)
                    }
                }
                try {
                    controller.invoke(callback, devicePtr, element, 0, DEFAULT_TIMEOUT_MS)
                } catch (e: Exception) {
                    onFailure("INVOKE_FAILED", e.message ?: "invoke threw", e)
                }
            },
            onFailure = { e -> onFailure("CONNECT_FAILED", e.message ?: "connect failed", e) }
        )
    }

    // -----------------------------------------------------------------------
    // SUBSCRIBE / UNSUBSCRIBE
    // -----------------------------------------------------------------------

    fun subscribe(
        matterNodeIdStr: String,
        attributePaths: ReadableArray?,
        minIntervalSec: Int,
        maxIntervalSec: Int,
        promise: Promise
    ) {
        val nodeId = parseMatterNodeId(matterNodeIdStr) ?: run {
            promise.reject("INVALID_ARG", "Invalid matterNodeId"); return
        }
        val controller = runCatching { ensureChipClient().chipDeviceController }
            .getOrElse { promise.reject("NO_FABRIC", it.message, it); return }

        // Build subscription paths. JS may omit `attributeId` to request a
        // *cluster-wildcard* subscription — needed for FW that rejects
        // explicit attribute paths but serialises the same attributes
        // under wildcard expansion (Google Home / chip-tool style).
        val paths = mutableListOf<ChipAttributePath>()
        if (attributePaths != null) {
            for (i in 0 until attributePaths.size()) {
                val m = attributePaths.getMap(i) ?: continue
                val endpointId = ChipPathId.forId(m.getInt("endpoint").toLong())
                val clusterId = ChipPathId.forId(m.getDouble("clusterId").toLong())
                val attributeId =
                    if (m.hasKey("attributeId") && !m.isNull("attributeId")) {
                        ChipPathId.forId(m.getDouble("attributeId").toLong())
                    } else {
                        ChipPathId.forWildcard()
                    }
                paths += ChipAttributePath.newInstance(endpointId, clusterId, attributeId)
            }
        }
        if (paths.isEmpty()) {
            promise.reject("INVALID_ARG", "subscribe: no attribute paths"); return
        }

        val handle = "sub-${subscriptionHandleSeq.incrementAndGet()}"
        val entry = SubscriptionEntry(handle, nodeId)
        subscriptions[handle] = entry

        val minInt = if (minIntervalSec > 0) minIntervalSec else DEFAULT_MIN_INTERVAL_SEC
        val maxInt = if (maxIntervalSec > 0) maxIntervalSec else DEFAULT_MAX_INTERVAL_SEC

        getConnectedDevicePointer(nodeId,
            onSuccess = { devicePtr ->
                val subEstablished = SubscriptionEstablishedCallback { subId ->
                    entry.subscriptionId = subId
                    Log.d(TAG, "Subscription $handle established: matterId=$subId")
                }
                val resub = ResubscriptionAttemptCallback { cause, nextMs ->
                    Log.d(TAG, "Subscription $handle re-attempt: cause=$cause, next=${nextMs}ms")
                }
                val report = object : ReportCallback {
                    override fun onReport(nodeState: NodeState?) {
                        if (nodeState == null) return
                        // Walk every endpoint/cluster/attribute the device
                        // reported. This must NOT iterate the requested
                        // `paths` because wildcard-subscribed clusters
                        // (e.g. PowerSource on RVC firmwares that quirk
                        // on concrete-path subscribes) expand on-device
                        // and the resulting attribute ids aren't known
                        // to the requestor.
                        for ((endpointBoxed, endpointState) in nodeState.endpointStates) {
                            val ep = endpointBoxed
                            for ((clusterBoxed, clusterState) in endpointState.clusterStates) {
                                val clu = clusterBoxed
                                for ((attributeBoxed, attributeState) in clusterState.attributeStates) {
                                    val att = attributeBoxed
                                    val raw = attributeState.value
                                    Log.d(
                                        TAG,
                                        "sub[$handle] report ep=$ep clu=0x${clu.toString(16)} " +
                                            "att=0x${att.toString(16)} raw=${MatterDataValueCodec.previewValue(raw)}"
                                    )
                                    val jsValue = MatterDataValueCodec.attributeValueToJs(raw)
                                    emitAttributeReport(matterNodeIdStr, ep, clu, att, jsValue)
                                }
                            }
                        }
                    }
                    override fun onError(
                        attributePath: ChipAttributePath?,
                        eventPath: ChipEventPath?,
                        ex: Exception
                    ) {
                        Log.w(TAG, "Subscription $handle report error: ${ex.message}")
                    }
                }

                try {
                    controller.subscribeToPath(
                        subEstablished,
                        resub,
                        report,
                        devicePtr,
                        paths,
                        emptyList<ChipEventPath>(),
                        minInt,
                        maxInt,
                        true,
                        false,
                        0
                    )
                    val res = Arguments.createMap().apply { putString("subscriptionId", handle) }
                    promise.resolve(res)
                } catch (e: Exception) {
                    subscriptions.remove(handle)
                    promise.reject("SUBSCRIBE_FAILED", e.message, e)
                }
            },
            onFailure = { e ->
                subscriptions.remove(handle)
                promise.reject("CONNECT_FAILED", e.message, e)
            }
        )
    }

    fun unsubscribe(handle: String, promise: Promise) {
        val entry = subscriptions.remove(handle)
        if (entry == null) {
            promise.resolve(successResult()); return
        }
        try {
            shutdownSubscriptionEntry(entry)
            promise.resolve(successResult())
        } catch (e: Exception) {
            promise.reject("UNSUBSCRIBE_FAILED", e.message, e)
        }
    }

    private fun shutdownSubscriptionEntry(entry: SubscriptionEntry) {
        val controller = runCatching { ensureChipClient().chipDeviceController }.getOrNull()
            ?: return
        try {
            // CHIP exposes overloads at multiple specificities; the
            // (fabricIndex, peerNodeId) form tears down every active
            // subscription for that node, matching our 1-handle-per-node usage.
            val fabricIndex = runCatching {
                ChipDeviceController::class.java
                    .getMethod("getFabricIndex")
                    .invoke(controller) as? Int ?: 0
            }.getOrDefault(0)

            try {
                ChipDeviceController::class.java.getMethod(
                    "shutdownSubscriptions",
                    Int::class.javaPrimitiveType,
                    Long::class.javaPrimitiveType
                ).invoke(controller, fabricIndex, entry.matterNodeId)
            } catch (e: NoSuchMethodException) {
                // Fall back to the no-arg overload if per-peer is unavailable.
                controller.shutdownSubscriptions()
            }
        } catch (e: Exception) {
            Log.w(TAG, "shutdownSubscriptions failed: ${e.message}")
        }
    }

    // -----------------------------------------------------------------------
    // INTERNALS
    // -----------------------------------------------------------------------

    private fun emitAttributeReport(
        matterNodeIdStr: String,
        endpoint: Int,
        clusterId: Long,
        attributeId: Long,
        jsValue: Any?
    ) {
        val event: WritableMap = Arguments.createMap().apply {
            putString("matterNodeId", matterNodeIdStr)
            putInt("endpoint", endpoint)
            putDouble("clusterId", clusterId.toDouble())
            putDouble("attributeId", attributeId.toDouble())
            putAnyValue("value", jsValue)
        }
        // RN bridge events must be dispatched on a thread with a Looper;
        // the CHIP report callback runs on the controller's worker thread.
        mainHandler.post {
            try {
                reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit(EVENT_ATTRIBUTE_REPORT, event)
            } catch (e: Exception) {
                Log.w(TAG, "emit attributeReport failed: ${e.message}")
            }
        }
    }

    private fun getConnectedDevicePointer(
        nodeId: Long,
        onSuccess: (Long) -> Unit,
        onFailure: (Exception) -> Unit
    ) {
        devicePtrs[nodeId]?.let { onSuccess(it); return }
        val controller = runCatching { ensureChipClient().chipDeviceController }
            .getOrElse { onFailure(it as? Exception ?: Exception(it.message)); return }
        try {
            controller.getConnectedDevicePointer(
                nodeId,
                object : GetConnectedDeviceCallback {
                    override fun onDeviceConnected(devicePointer: Long) {
                        devicePtrs[nodeId] = devicePointer
                        onSuccess(devicePointer)
                    }
                    override fun onConnectionFailure(failedNodeId: Long, error: Exception?) {
                        onFailure(error ?: Exception("connection failure for $failedNodeId"))
                    }
                }
            )
        } catch (e: Exception) {
            onFailure(e)
        }
    }

    /** Reuse the active commissioning ChipClient or lazily reconstruct from FabricInfo. */
    private fun ensureChipClient(): ChipClient {
        val existing = FabricSessionManager.getCurrentChipClient()
        if (existing != null) return existing

        val resolved = FabricSessionManager.resolveChipClient(
            reactContext.applicationContext as Context
        )
        if (resolved != null) return resolved

        throw IllegalStateException("No active fabric for Matter control")
    }

    private fun parseMatterNodeId(s: String?): Long? {
        if (s.isNullOrBlank()) return null
        val cleaned = s.removePrefix("0x").removePrefix("0X")
        return runCatching { BigInteger(cleaned, 16).toLong() }
            .recoverCatching { java.lang.Long.parseUnsignedLong(s) }
            .getOrNull()
    }

    private fun successResult(): WritableMap =
        Arguments.createMap().apply { putBoolean("success", true) }
}

/** Helper to put a heterogeneous JS value into a [WritableMap] without cluttering call sites. */
private fun WritableMap.putAnyValue(key: String, value: Any?) {
    when (value) {
        null -> putNull(key)
        is Boolean -> putBoolean(key, value)
        is Int -> putInt(key, value)
        is Double -> putDouble(key, value)
        is String -> putString(key, value)
        is com.facebook.react.bridge.WritableArray -> putArray(key, value)
        is com.facebook.react.bridge.WritableMap -> putMap(key, value)
        else -> putString(key, value.toString())
    }
}

File: ios/Matter/ESPMatterControl.swift
/*
 * SPDX-FileCopyrightText: 2026 Espressif Systems (Shanghai) CO LTD
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation
import Matter
import React

/// Native Matter control adapter for iOS. Mirrors the Android
/// `ESPMatterControl` surface and exposes the four canonical Matter
/// Interaction Model operations — `read`, `write`, `invoke`,
/// `subscribe` (plus their lifecycle siblings `init`, `shutdown`,
/// `unsubscribe`).
///
/// Outbound `MatterDataValue` payloads are passed through
/// `normalizeOutboundDataValue` to the Apple `MTRDataValueDictionary`
/// shape that `MTRBaseDevice.invokeCommand…/writeAttribute…` expects.
/// Inbound subscription reports are reduced to a primitive JS-friendly
/// value and forwarded over the `ESPMatter:attributeReport` RN event —
/// same name as Android — so a single JS subscription channel can
/// listen on either platform.
///
/// No cluster-specific semantic translation lives here. Callers send
/// raw cluster / attribute / command ids and Matter data-value payloads;
/// semantic mapping (semantic units, OnOff bool, mode pickers, …) lives
/// above this surface in TypeScript hooks/panels or in the Matter SDK
/// transformer.
public enum ESPMatterControl {
  public static let attributeReportEventName: String = "ESPMatter:attributeReport"

  fileprivate static let DEFAULT_MIN_INTERVAL_SEC: UInt16 = 1
  fileprivate static let DEFAULT_MAX_INTERVAL_SEC: UInt16 = 30
}

extension ESPMatterModule {

  // MARK: - Lifecycle (init / shutdown)

  @objc func matterControlInit(_ config: NSDictionary?,
                               resolver resolve: @escaping RCTPromiseResolveBlock,
                               rejecter _: @escaping RCTPromiseRejectBlock) {
    // The active controller is set up by commissioning; init is a
    // no-op placeholder for symmetry with Android.
    _ = config
    resolve(["success": true])
  }

  @objc func matterControlShutdown(_ resolve: @escaping RCTPromiseResolveBlock,
                                   rejecter _: @escaping RCTPromiseRejectBlock) {
    // We do not own the controller's lifecycle here; commissioning
    // owns it. Drop any per-device report handlers we registered
    // and resolve.
    resolve(["success": true])
  }

  // MARK: - Read / Write / Invoke

  @objc func matterControlRead(_ matterNodeId: String,
                               endpoint: NSNumber,
                               clusterId: NSNumber,
                               attributeId: NSNumber,
                               resolver resolve: @escaping RCTPromiseResolveBlock,
                               rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let node = ESPMatterModule.parseMatterNodeId(matterNodeId),
          let controller = currentMatterController else {
      reject("NO_CONTROLLER", "Matter controller is not initialised", nil); return
    }
    let device = MTRBaseDevice(nodeID: node, controller: controller)
    device.readAttributes(withEndpointID: endpoint,
                          clusterID: clusterId,
                          attributeID: attributeId,
                          params: nil,
                          queue: matterQueue) { reports, error in
      if let error = error {
        reject("READ_FAILED", error.localizedDescription, error); return
      }
      let dataValue = ESPMatterModule.firstAttributeData(from: reports)
      let primitive = ESPMatterModule.dataValueToPrimitive(dataValue) ?? NSNull()
      resolve(["success": true, "value": primitive])
    }
  }

  @objc func matterControlWrite(_ matterNodeId: String,
                                endpoint: NSNumber,
                                clusterId: NSNumber,
                                attributeId: NSNumber,
                                value: NSDictionary?,
                                resolver resolve: @escaping RCTPromiseResolveBlock,
                                rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let node = ESPMatterModule.parseMatterNodeId(matterNodeId),
          let controller = currentMatterController else {
      reject("NO_CONTROLLER", "Matter controller is not initialised", nil); return
    }
    guard let value = value else {
      reject("INVALID_ARG", "write: value is nil", nil); return
    }
    let device = MTRBaseDevice(nodeID: node, controller: controller)
    let normalised = ESPMatterModule.normalizeOutboundDataValue(value)
    device.writeAttribute(withEndpointID: endpoint,
                          clusterID: clusterId,
                          attributeID: attributeId,
                          value: normalised,
                          timedWriteTimeout: nil,
                          queue: matterQueue) { _, error in
      if let error = error {
        reject("WRITE_FAILED", error.localizedDescription, error); return
      }
      resolve(["success": true])
    }
  }

  @objc func matterControlInvoke(_ matterNodeId: String,
                                 endpoint: NSNumber,
                                 clusterId: NSNumber,
                                 commandId: NSNumber,
                                 commandFields: NSDictionary?,
                                 resolver resolve: @escaping RCTPromiseResolveBlock,
                                 rejecter reject: @escaping RCTPromiseRejectBlock) {
    invokeInternal(matterNodeId: matterNodeId,
                   endpoint: endpoint,
                   clusterId: clusterId,
                   commandId: commandId,
                   commandFields: commandFields,
                   onSuccess: { resolve(["success": true]) },
                   onFailure: { code, message, err in reject(code, message, err) })
  }

  @objc func matterEncodeCommandFieldsToTlvHex(_ commandFields: NSDictionary?,
                                               resolver resolve: @escaping RCTPromiseResolveBlock,
                                               rejecter reject: @escaping RCTPromiseRejectBlock) {
    let hex = MatterDataValueCodec.encodeCommandFieldsToTlvHex(commandFields)
    resolve(hex)
  }

  // MARK: - Subscribe / Unsubscribe

  @objc func matterControlSubscribe(_ matterNodeId: String,
                                    attributePaths: NSArray?,
                                    minIntervalSec: NSNumber,
                                    maxIntervalSec: NSNumber,
                                    resolver resolve: @escaping RCTPromiseResolveBlock,
                                    rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let node = ESPMatterModule.parseMatterNodeId(matterNodeId),
          let controller = currentMatterController else {
      reject("NO_CONTROLLER", "Matter controller is not initialised", nil); return
    }
    guard let attributePaths = attributePaths, attributePaths.count > 0 else {
      reject("INVALID_ARG", "subscribe: no attribute paths", nil); return
    }

    let device = MTRBaseDevice(nodeID: node, controller: controller)
    let minRequested = minIntervalSec.uint16Value
    let maxRequested = maxIntervalSec.uint16Value
    let minInt = max(ESPMatterControl.DEFAULT_MIN_INTERVAL_SEC, minRequested)
    let maxInt = max(UInt16(minInt + 1),
                     maxRequested > 0 ? maxRequested : ESPMatterControl.DEFAULT_MAX_INTERVAL_SEC)

    let params = MTRSubscribeParams(minInterval: NSNumber(value: minInt),
                                    maxInterval: NSNumber(value: maxInt))
    params.shouldResubscribeAutomatically = true

    var allow = Set<UInt64>()
    for case let path as NSDictionary in attributePaths {
      guard let ep = (path["endpoint"] as? NSNumber),
            let cl = (path["clusterId"] as? NSNumber),
            let at = (path["attributeId"] as? NSNumber) else { continue }
      allow.insert(ESPMatterModule.packPath(endpoint: ep.intValue,
                                            clusterId: cl.uint64Value,
                                            attributeId: at.uint64Value))
    }
    if allow.isEmpty {
      reject("INVALID_ARG", "subscribe: no valid attribute paths", nil); return
    }

    let handle = "sub-\(UUID().uuidString)"
    let listenerNodeId = matterNodeId

    device.subscribeToAttributes(
      withEndpointID: nil,
      clusterID: nil,
      attributeID: nil,
      params: params,
      queue: matterQueue,
      reportHandler: { [weak self] reports, _ in
        guard let self = self else { return }
        self.handleSubscriptionReports(matterNodeId: listenerNodeId,
                                       reports: reports,
                                       allow: allow)
      },
      subscriptionEstablished: {}
    )

    resolve(["subscriptionId": handle])
  }

  @objc func matterControlUnsubscribe(_ subscriptionId: String,
                                      resolver resolve: @escaping RCTPromiseResolveBlock,
                                      rejecter _: @escaping RCTPromiseRejectBlock) {
    // `MTRBaseDevice` does not expose per-handle teardown. The standard
    // Apple pattern is `deregisterReportHandlers` per device, which we
    // fold into `matterControlShutdown`. We accept the handle for API
    // parity with Android and resolve as a no-op.
    _ = subscriptionId
    resolve(["success": true])
  }

  // MARK: - Internals

  private func invokeInternal(matterNodeId: String,
                              endpoint: NSNumber,
                              clusterId: NSNumber,
                              commandId: NSNumber,
                              commandFields: NSDictionary?,
                              onSuccess: @escaping () -> Void,
                              onFailure: @escaping (String, String, Error?) -> Void) {
    guard let node = ESPMatterModule.parseMatterNodeId(matterNodeId),
          let controller = currentMatterController else {
      onFailure("NO_CONTROLLER", "Matter controller is not initialised", nil); return
    }
    let device = MTRBaseDevice(nodeID: node, controller: controller)
    // Apple requires `commandFields` to be a Structure data-value, even
    // for fieldless commands. Default to an empty Structure when JS
    // sends nil.
    let fields: [String: Any] = commandFields.map {
      ESPMatterModule.normalizeOutboundDataValue($0)
    } ?? ["type": "Structure", "value": [[String: Any]]()]

    device.invokeCommand(withEndpointID: endpoint,
                         clusterID: clusterId,
                         commandID: commandId,
                         commandFields: fields,
                         timedInvokeTimeout: nil,
                         queue: matterQueue) { _, error in
      if let error = error {
        onFailure("INVOKE_FAILED", error.localizedDescription, error); return
      }
      onSuccess()
    }
  }

  fileprivate func handleSubscriptionReports(matterNodeId: String,
                                             reports: [Any]?,
                                             allow: Set<UInt64>? = nil) {
    guard let reports = reports as? [[String: Any]] else { return }
    for report in reports {
      guard let path = report["attributePath"] as? MTRAttributePath else { continue }
      if let allow = allow {
        let key = ESPMatterModule.packPath(endpoint: path.endpoint.intValue,
                                           clusterId: path.cluster.uint64Value,
                                           attributeId: path.attribute.uint64Value)
        if !allow.contains(key) { continue }
      }
      let value = report["data"] as? [String: Any]
      let primitive = ESPMatterModule.dataValueToPrimitive(value) ?? NSNull()
      let event: [String: Any] = [
        "matterNodeId": matterNodeId,
        "endpoint": path.endpoint.intValue,
        "clusterId": path.cluster.doubleValue,
        "attributeId": path.attribute.doubleValue,
        "value": primitive
      ]
      sendEvent(withName: ESPMatterControl.attributeReportEventName, body: event)
    }
  }

  // MARK: - Helpers (static so the extension can stay stateless)

  fileprivate static func parseMatterNodeId(_ s: String) -> NSNumber? {
    if s.isEmpty { return nil }
    let cleaned = s.hasPrefix("0x") || s.hasPrefix("0X") ? String(s.dropFirst(2)) : s
    if let v = UInt64(cleaned, radix: 16) { return NSNumber(value: v) }
    if let v = UInt64(s) { return NSNumber(value: v) }
    return nil
  }

  fileprivate static func packPath(endpoint: Int,
                                   clusterId: UInt64,
                                   attributeId: UInt64) -> UInt64 {
    let ep = UInt64(endpoint & 0xFFFF)
    let cl = clusterId & 0xFF_FFFF
    let at = attributeId & 0xFF_FFFF
    return (ep << 48) | (cl << 24) | at
  }

  // MARK: - MatterDataValue ↔ Apple data-value-dict bridging
  //
  // The JS wire format mirrors Apple's `MTRDataValueDictionary` keys
  // exactly (`type`/`value`/`contextTag`/`data`), so the bridge is
  // mostly the identity. We only:
  //   - decode `OctetString.value` from base64 string → `NSData`
  //   - recurse into Structure / Array entries to apply the same rule
  // The output dictionary is what `MTRBaseDevice.invokeCommand…
  // commandFields:` and `writeAttribute… value:` expect.

  fileprivate static func normalizeOutboundDataValue(_ raw: NSDictionary) -> [String: Any] {
    guard let type = raw["type"] as? String else {
      return raw as? [String: Any] ?? [:]
    }
    var out: [String: Any] = ["type": type]
    switch type {
    case "OctetString":
      if let s = raw["value"] as? String,
         let data = Data(base64Encoded: s) {
        out["value"] = data
      } else if let data = raw["value"] as? Data {
        out["value"] = data
      }
    case "Structure":
      let entries = (raw["value"] as? [[String: Any]]) ?? []
      out["value"] = entries.map { entry -> [String: Any] in
        var e: [String: Any] = [:]
        if let tag = entry["contextTag"] as? NSNumber { e["contextTag"] = tag }
        if let inner = entry["data"] as? NSDictionary {
          e["data"] = normalizeOutboundDataValue(inner)
        }
        return e
      }
    case "Array":
      let entries = (raw["value"] as? [[String: Any]]) ?? []
      out["value"] = entries.map { entry -> [String: Any] in
        var e: [String: Any] = [:]
        if let inner = entry["data"] as? NSDictionary {
          e["data"] = normalizeOutboundDataValue(inner)
        }
        return e
      }
    case "Null":
      // Omit `value` per spec.
      break
    default:
      // Pass primitives through verbatim.
      if let v = raw["value"] { out["value"] = v }
    }
    return out
  }

  /// Pull the first attribute's data-value out of a heterogeneous
  /// report array as returned by `readAttributes`/`subscribeToAttributes`.
  fileprivate static func firstAttributeData(from reports: [Any]?) -> [String: Any]? {
    guard let reports = reports as? [[String: Any]] else { return nil }
    return reports.first?["data"] as? [String: Any]
  }

  /// Flatten an Apple-data-value-dict to a primitive JS-compatible value.
  fileprivate static func dataValueToPrimitive(_ value: [String: Any]?) -> Any? {
    guard let value = value, let type = value["type"] as? String else { return nil }
    switch type {
    case "Null":
      return NSNull()
    case "Boolean":
      return value["value"] as? Bool ?? false
    case "UnsignedInteger", "SignedInteger", "Float", "Double":
      return value["value"] as? NSNumber
    case "UTF8String":
      return value["value"] as? String
    case "OctetString":
      if let d = value["value"] as? Data { return d.base64EncodedString() }
      return value["value"]
    case "Structure":
      let entries = (value["value"] as? [[String: Any]]) ?? []
      var dict: [String: Any] = [:]
      for entry in entries {
        guard let tag = entry["contextTag"] as? NSNumber,
              let inner = entry["data"] as? [String: Any] else { continue }
        dict["\(tag.intValue)"] = dataValueToPrimitive(inner) ?? NSNull()
      }
      return dict
    case "Array":
      let entries = (value["value"] as? [[String: Any]]) ?? []
      return entries.compactMap { e -> Any? in
        (e["data"] as? [String: Any]).flatMap(dataValueToPrimitive)
      }
    default:
      return value["value"]
    }
  }
}
